src/main/java/io/cdap/plugin/sap/SapODataSource.java

```diff
@@ -129,86 +152,322 @@ public Schema getSchema() {
+        // Olingo V4 client does not parse expression annotations correctly:
+        // https://issues.apache.org/jira/browse/OLINGO-1403
+        CsdlLabeledElement labeledElement = expression.asDynamic().asLabeledElement();
+        Schema labeledElementSchema = expressionToFieldSchema(fieldName, labeledElement.getValue());
+        return SapODataConstants.LabeledElementExpression.schema(fieldName, labeledElementSchema);
+      case Record:
+        CsdlRecord record = expression.asDynamic().asRecord();
+        Schema propertyValuesSchema = propertyValuesSchema(fieldName, record.getPropertyValues());
+        if (record.getAnnotations() == null || record.getAnnotations().isEmpty()) {
+          return SapODataConstants.RecordExpression.schema(fieldName, propertyValuesSchema, null);
+        }
+        Map<String, OData4Annotation> nestedRecordAnnotations = record.getAnnotations().stream()
+          .map(OData4Annotation::new)
+          .collect(Collectors.toMap(OData4Annotation::getName, Function.identity()));
+        Schema recordAnnotationsSchema = nestedAnnotationsSchema(recordName, nestedRecordAnnotations);
+        return SapODataConstants.RecordExpression.schema(fieldName, propertyValuesSchema, recordAnnotationsSchema);
+      case UrlRef:
+        CsdlUrlRef urlRef = expression.asDynamic().asUrlRef();
+        Schema urlRefValueSchema = expressionToFieldSchema(fieldName, urlRef.getValue());
+        return SapODataConstants.UrlRefExpression.schema(fieldName, urlRefValueSchema);
+      default:
+        // this should never happen
+        throw new InvalidStageException(
+          String.format("Annotation expression for field '%s' is of unsupported type '%s'.", fieldName, type));
+    }
+  }
+
+  private Schema propertyValuesSchema(String fieldName, List<CsdlPropertyValue> propertyValues) {
+    if (propertyValues == null || propertyValues.isEmpty()) {
+      return null;
+    }
+    List<Schema.Field> fields = propertyValues.stream()
+      .map(p -> Schema.Field.of(p.getProperty(), expressionToFieldSchema(fieldName, p.getValue())))
+      .collect(Collectors.toList());
+
+    return Schema.recordOf(fieldName + "-property-values", fields);
+  }
+
+  private Schema propertyToSchema(PropertyMetadata propertyMetadata) {
+    Schema schema = propertyTypeToSchema(propertyMetadata);
+    return propertyMetadata.isCollection() ? Schema.arrayOf(schema) : schema;
   }
 
-  private Schema convertPropertyType(PropertyMetadata propertyMetadata) {
-    switch (propertyMetadata.getEdmTypeName()) {
+  private Schema propertyTypeToSchema(PropertyMetadata propertyMetadata) {
+    if (propertyMetadata.isComplex()) {
+      return convertComplexProperty(propertyMetadata.getName(), (ComplexPropertyMetadata) propertyMetadata);
+    }
+    if (propertyMetadata.isEnum()) {
+      return Schema.of(Schema.Type.STRING);
+    }
+    switch (propertyMetadata.getType()) {
       case "Binary":
+      case "Edm.Binary":
```

Comment:
how come there are copies of each of these types?

Reply:
OData 4 client changed to use XML metadata request instead of metadata request since it's not possible to fetch property annotations in the case of missing vocabulary references: https://issues.apache.org/jira/browse/OLINGO-1130
I faced this issue while testing the plugin with public OData V4 services.

So, currently, we use other model classes that use the "Edm." prefix for OData V4 data types.